Fix age-restricted videos

xBaank opened the pull request:
<!--

⚠ Note this project is in maintenance mode and no longer accepts pull requests for new features.
Pull requests for bug fixes are still welcome, however.

-->

<!--

**Important:**

Please make sure that there is an existing issue that describes the problem solved by your pull request. If there isn't one, consider creating it first.
An open issue offers a good place to iron out requirements, discuss possible solutions, and ask questions.

Remember to also:

- Keep your pull request focused and as small as possible. If you want to contribute multiple unrelated changes, please create separate pull requests for them.
- Follow the coding style and conventions already established by the project. When in doubt about which style to use, ask in the comments to your pull request.
- If you want to start a discussion regarding a specific change you've made, add a review comment to your own code. This can be used to highlight something important or to seek further input from others.

-->

<!-- Please specify the issue addressed by this pull request -->
Closes #581

I also changed one small thing in the tests because tests were failing (your channel now have 78 videos instead of 79).

Another thing was that videos that requires purchase could not throw an exception if purchasePreviewVideoId was not found, so I added a method to check if the reason of not being playable was "This video requires payment to watch.".

Tyrrrz commented:
Hi! Thanks a lot for the PR. Can you add `.vs` to `.gitignore` please?

Tyrrrz commented on YoutubeExplode/Bridge/YoutubeController.cs:
```diff
@@ -124,6 +124,23 @@ internal class YoutubeController
 
+        public async Task<PlayerResponseExtractor?> GetPlayerResponseFromEndpoint(VideoId videoId)
+        {
+            string url = "https://www.youtube.com/youtubei/v1/player";
+
+            //Use this as mentioned on https://github.com/Tyrrrz/YoutubeExplode/issues/581#issuecomment-889241520
+            string requestJson = $@"{{""context"": {{""client"": {{""clientName"": ""ANDROID"",""clientScreen"": ""EMBED"",""clientVersion"": ""16.05""}},""thirdParty"": {{""embedUrl"": ""https://www.youtube.com""}}}},""videoId"": ""{videoId}"" }}";
+
+
+            HttpRequestMessage httpRequestMessage = new HttpRequestMessage(HttpMethod.Post,url);
+            httpRequestMessage.Headers.Add("X-Goog-Api-Key", InternalApiKey);
+            httpRequestMessage.Content = new StringContent(requestJson, System.Text.Encoding.UTF8, "application/json");
```
Can we rewrite it to use object serialization, similar to `GetPlaylistAsync(...)` method? The string is a bit hard to read.

https://github.com/Tyrrrz/YoutubeExplode/blob/314be1cade649064ba8f3e6620b812a4ea651539/YoutubeExplode/Bridge/YoutubeController.cs#L160-L187

Tyrrrz commented on YoutubeExplode/Videos/Streams/StreamClient.cs:
```diff
@@ -194,17 +195,22 @@ public StreamClient(HttpClient httpClient)
+                    playerResponseFromWatchPage = await _controller.GetPlayerResponseFromEndpoint(videoId) ?? playerResponseFromWatchPage;
+
                 var purchasePreviewVideoId = playerResponseFromWatchPage.TryGetPreviewVideoId();
-                if (!string.IsNullOrWhiteSpace(purchasePreviewVideoId))
+                bool isPurchaseRequired = playerResponseFromWatchPage.TryGetPurchaseRequired();
```
Do we need `isPurchaseRequired` in addition to `purchasePreviewVideoId`? It seems if the latter is not null, it's fair to assume that the video is age-restricted.

Tyrrrz commented on YoutubeExplode/Bridge/Extractors/VideoWatchPageExtractor.cs:
```diff
@@ -89,6 +88,8 @@ internal partial class VideoWatchPageExtractor
                 .Pipe(j => new PlayerResponseExtractor(j))
         );
 
+        public static PlayerResponseExtractor? TryGetPlayerResponse(string json) => String.IsNullOrWhiteSpace(json) ? null : new PlayerResponseExtractor(Json.Parse(json));
```
Can `json` ever be null or whitespace here? It seems to be passed from an HTTP request where it's reasonable to assume it's always going to be non-whitespace.

Tyrrrz commented:
Also: looks like the build failed on CI

xBaank commented:
I removed isPurchaseRequired as is not needed (I don't know why i added it, because now it's finding the previewVideoId), changed it to use object serialization, removed IsNullOrWhiteSpace of the json and other things so it passes the build.

Tyrrrz commented on YoutubeExplode/Bridge/Extractors/VideoWatchPageExtractor.cs:
```diff
@@ -127,6 +128,9 @@ internal partial class VideoWatchPageExtractor
         {
             var content = Html.Parse(raw);
 
+            if (content is null)
+                return null;
+
             var isValid = content.Body.QuerySelector("#player") is not null;
```
```suggestion
            var isValid = content?.Body?.QuerySelector("#player") is not null;
```

Tyrrrz commented on YoutubeExplode/Bridge/Extractors/VideoWatchPageExtractor.cs:
```diff
@@ -127,7 +128,7 @@ internal partial class VideoWatchPageExtractor
             var content = Html.Parse(raw);
 
-            var isValid = content.Body.QuerySelector("#player") is not null;
+            var isValid = content?.Body?.QuerySelector("#player") is not null;
```
```suggestion
            var isValid = content.Body?.QuerySelector("#player") is not null;
```
